python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1008,6 +1028,207 @@ predicate subscriptReadStep(CfgNode nodeFrom, Content c, CfgNode nodeTo) {
+    }
+
+    Expr getValue() { result = value }
+  }
+
+  /** A (possibly recursive) target of an unpacking assignment */
+  class UnpackingAssignmentTarget extends ControlFlowNode {
+    UnpackingAssignmentTarget() {
+      this instanceof UnpackingAssignmentDirectTarget
+      or
+      this = any(UnpackingAssignmentTarget parent).getAnElement()
+    }
+
+    ControlFlowNode getElement(int i) { result = this.(SequenceNode).getElement(i) }
+
+    ControlFlowNode getAnElement() { result = this.getElement(_) }
+  }
+
+  predicate unpackingAssignmentFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target | target instanceof SequenceNode |
+      nodeFrom = TIterableSequence(target) and
+      nodeTo.asCfgNode() = target
+    )
+  }
+
+  predicate unpackingAssignmentConvertingReadStep(Node nodeFrom, Content c, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target | target instanceof SequenceNode |
+      nodeFrom = TIterableSequence(target) and
+      nodeTo = TIterableElement(target) and
+      (
+        c instanceof ListElementContent
+        or
+        c instanceof SetElementContent
+        or
+        // do not lose precision by routing tuple content through the `IterableElement`
+        not target instanceof TupleNode and
+        // `index` refers to `nodeFrom`, but only the ones in `target` are relevant.
+        exists(int index | exists(target.getElement(index)) |
+          c.(TupleElementContent).getIndex() = index
+        )
+        // leaving out dict content for now
+      )
+    )
+  }
+
+  predicate unpackingAssignmentConvertingStoreStep(Node nodeFrom, Content c, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target | target instanceof SequenceNode |
+      nodeFrom = TIterableElement(target) and
+      nodeTo.asCfgNode() = target and
+      (
+        target instanceof ListNode and
+        c instanceof ListElementContent
+        or
+        target instanceof TupleNode and
+        exists(int index | exists(target.getElement(index)) |
+          c.(TupleElementContent).getIndex() = index
+        )
+      )
+    )
+  }
+
+  predicate unpackingAssignmentElementReadStep(Node nodeFrom, Content c, Node nodeTo) {
+    exists(UnpackingAssignmentTarget target, int index, ControlFlowNode element, boolean precise |
+      target instanceof SequenceNode
+    |
+      nodeFrom.asCfgNode() = target and
+      element = target.getElement(index) and
+      (
+        target instanceof ListNode and
+        c instanceof ListElementContent
+        or
+        target instanceof TupleNode and
+        if precise = true
+        then c.(TupleElementContent).getIndex() = index
+        else c instanceof TupleElementContent // This could get big if big tuples exist
+      ) and
+      (
+        if element instanceof SequenceNode
+        then
+          nodeTo = TIterableSequence(element) and
+          precise = true
+        else
+          if element.getNode() instanceof Starred
```

Comment:
Looking at this and comparing it with line 1191, I was wondering if these could actually overlap (and so whether a simple  disjunction would suffice). Then I realised that the following abomination is valid:
```py
a, *(b, c) = (1,2,3)
```
Do you know if we handle the above correctly?

Reply:
yikes :smile:

Reply:
I would expect that we do not...

Reply:
Test added, we do indeed not.